### DISCUSSION TITLE
Signature generálás

### DISCUSSION THREAD

**Discussion by czeba-b** [nav-gov-hu/M2M#37] [Kérdések] [answered]
Sziasztok!
Az Activation endpoint hívásánál van gondom a signature generálással, de leírom a folyamatomat hátha máshol van a hiba.
Teszt API-t használom.
#### 1. /NavM2mCommon/tokenService/Token
Token igénylésénél nincs gond, az szépen megjön.
#### 2. /NavM2mCommon/userregistrationService/Nonce
A kérést le tudom futtatni eredményesen, de a dokumentációktól eltérően.

A common swagger fájlban az van hogy a RedeemNonceRequestType-ba "Az aláírókulcs második felének elkérésére alkalmas ideiglenes jelszó" kell.
Nálam itt csak az "Aláírókulcs első fele" kódot fogadja el, az "Aláírókulcs második felének kiváltásához szükséges kód" megadásakor az jön vissza hogy:
```
{'resultMessage': 'Nonce is empty', 'signatureKeySecondPart': None, 'resultCode': 'INVALID_NONCE'}
```
Ellenőriztem többször, nem üres a kérésben a nonce.

Az "Aláírókulcs első fele" használatával ez jön ami jónak tűnik:
```
{'resultMessage': '', 'signatureKeySecondPart': 'xxxxx', 'resultCode': 'REDEEM_NONCE_SUCCESSFUL'}
```
Akárhányszor ismétlem meg a lekérést, mindig ugyanaz a signatureKeySecondPart kód jön.
Ez így eddig rendben van?
### 3. /NavM2mCommon/userregistrationService/Activation
Itt van leginkább gondom, nem akarja elfogadni a signature-t.
Ilyen módon állítom össze:
`UPPER(BASE64(SHA256(message_id+timestamp+signature_key_first_part+signature_key_second_part)))`
Ahol:
- message_id: ugyanaz mint a Token és Nonce kéréseknél volt
- timestamp: UTC időbélyeg `"%Y%m%d%H%M%S"` formátumban (Pl: `20241114114326`)
- signature_key_first_part: Aláírókulcs első fele
- signature_key_second_part: nonce lekérdezésből `signatureKeySecondPart`

Kipróbáltam, ugyanazt a hash-t számolom mint amit a postman-es pre-request script.

Példa request CURL formátumban:
```sh
curl -X POST -H 'Accept: */*' -H 'Accept-Encoding: gzip, deflate' \
    -H 'Authorization: Bearer AAAAAAAAAAAAAD9gGDsIYH4hfj4DPQdPK...' \
    -H 'Connection: keep-alive' -H 'Content-Length: 78' \
    -H 'Content-Type: application/json' \
    -H 'messageId: e83b2c7b-0bf2-4eaf-babc-8cc4d44850c9' \
    -d '{"requestData": {"signature": "7QHIFBLTTNM2ZP0TKR1IWH0QOIPJHOOUNY1RUNGTMTM="}}' \
    https://m2m-dev.nav.gov.hu/rest-api/1.1/NavM2mCommon/userregistrationService/Activation
```
Ez a válasz jön:
```
{'resultMessage': '', 'resultCode': 'INVALID_SIGNATURE'}
```

Mi lehet a gond?
Példa kérés ma reggelről:
message_id: `e83b2c7b-0bf2-4eaf-babc-8cc4d44850c9`

Általános kérdés: SHA-256 kell vagy SHA3-256? Vagy van ahol az egyik van ahol a másik?
Vegyesen van írva a postman-ben és a swagger fájlokban, van ahol ellentmondás is van, pl:
- Fájlfeltöltés
	- Postman: 
		```
		var toSign = message_id+timestamp+file_hash+signature_key_first_part+signature_key_second_part;
		console.log(toSign);
		var signature = cryptoJs.SHA256(toSign).toString(cryptoJs.enc.Base64).toUpperCase();
        ```
	- Swagger (Itt a hashmód is más és a hashképzés elemei is, bár gondolom a signatureKey az  a signature_key_first_part+signature_key_second_part akar lenni):
		```
		- name: signature
		  in: query
		  description: >-
			Fájlfeltöltés aláírása. Aláírás képzése:
			SHA3-256(messageId+timestamp+sha256hash+signatureKey) 
		  schema:
			type: string
		```

**Comment by TamasPill** [comment, marked as answer]
**Nonce beváltás:**
"Akárhányszor ismétlem meg a lekérést, mindig ugyanaz a signatureKeySecondPart kód jön.
Ez így eddig rendben van?"
Igen, ez így helyes. 
Éles környezeten a felhasználó aktiválás után már nem fog többet választ adni a nonce beváltás az adott nonce-ra.
Teszt környezeten az aktiválás után is mindig visszaadja a választ, hogy tesztelhető legyen az API.

Az ÜPO Tesztkörnyezeti adatok megjelenítésénél valóban fel van cserélve a két mező adattartalma, ezt javítani fogjuk hamarosan.


**Felhasználó aktiválás**
SHA256-ot kell használni minden esetben, ez most javításra került a swagger kommentben.
Az aláírás képzés képlete jól van leírva a fenti kommentben, így nem látszik, hogy melyik részében van a probléma.
Egy friss tesztkörnyezet regisztrációval a Postman csomagból próbálva sikeres az aktiválás, ezért a generálásnál lehet valami gond még.

**Reply by czeba-b** [reply]
Köszönöm a válaszokat!

Itt volt a kutya elásva:
> Az ÜPO Tesztkörnyezeti adatok megjelenítésénél valóban fel van cserélve a két mező adattartalma, ezt javítani fogjuk hamarosan.

A felcserélés miatt én is felcserélve állítottam össze a signature adattartalmát. Megcseréltem a kettőt, így már rendben van.
```
{'resultMessage': '', 'resultCode': 'ACTIVATE_USER_REGISTRATION_SUCCESSFUL'}
```

**Reply by e-lias** [reply]
Üdv! Becsatlakoznék ide, hogy ontopic legyek. Nálam is INVALID_SIGNATURE válasz van most. Elolvastam az itteni problémát, de nem igazán értem, hogy melyik mezőket kellett felcserélni, hogy működjön.

Én így állítom össze az aláírást:
`$act_messageId = generateMessageId();`
`$endusersecret_string = $act_messageId.date("YmdHis").$alairokulcs_elsofele.$kulcs_madosikresz;`
`$endusersecret = strtoupper(base64_encode(hash("sha256", $endusersecret_string)));`

Előre is köszönöm!

**Reply by TamasPill** [reply]
Az itteni probléma már megoldódott, így jól jelennek meg az adatok az Ügyfélportálon.
A mezők sorrendje jó a scriptben, ha ugyanez a messageId van a kérésben is és a dátum formátum megfelelő és UTC szerinti, akkor jónak kellene lennie felhasználó aktiválásra.
A többi szolgáltatás, ami vár aláírást, azokban van további paraméter is, amiről a dokumentáció ír.

**Reply by e-lias** [reply]
User activation-ben kapom ezt.
A dátumformátum szerintem jó, bár nem utc volt, hanem local. De utc zónában is invalid signature-t ad.

MessageId: MSG04505960017441010402822600000
Dátum: 20250408083040

**Reply by TamasPill** [reply]
Logok alapján azt látom, hogy már történtek sikeres aktiválások a 31XFYq0pB0 userrel.

**Reply by e-lias** [reply]
Igen, bújtam itt a topicocat közben és az egyik javaslat megoldotta. Köszönöm!

**Reply by Ficsy** [reply]
> Igen, bújtam itt a topicocat közben és az egyik javaslat megoldotta. Köszönöm!

Megírnád kérlek, hogy mi oldotta meg?  Én is itt vagyok elakadva...
